src/structures/paging/frame_alloc.rs

```diff
@@ -0,0 +1,15 @@
+//! Traits for abstracting away frame allocation and deallocation.
+
+use structures::paging::{PageSize, PhysFrame};
+
+/// A trait for types that can allocate a frame of memory.
+pub trait FrameAllocator {
+    /// Allocate a frame of the appropriate size and return it if possible.
+    fn alloc<S: PageSize>(&mut self) -> Option<PhysFrame<S>>;
+}
+
+/// A trait for types that can deallocate a frame of memory.
+pub trait FrameDeallocator {
+    /// Deallocate the given frame of memory.
+    fn dealloc<S: PageSize>(&mut self, frame: PhysFrame<S>);
```

Comment:
As above:

> We should move the generic paramenter `S` to the trait, otherwise every implementation must be able to allocate frames of all possible sizes (would this even be possible?).